vmprof/show.py

```diff
@@ -128,6 +127,82 @@ def print_node(parent, node, level):
         self._walk_tree(None, tree, 0, print_node)
 
 
+NodeDescr = namedtuple(
+        'NodeDescr',
+        ['block_type', 'funname', 'funline', 'filename'])
+
+
+def parse_block_name(node_name):
+    nparts = node_name.count(':')+1
+
+    block_type = None
+    funline = None
+    filename = None
+    if nparts == 4:
+        block_type, funname, funline, filename = node_name.split(':')
+    elif nparts == 2:
+        block_type, funname = node_name.split(':')
+    else:
+        funname = node_name
+
+    return NodeDescr(block_type, funname, funline, filename)
+
+
+class FlatPrinter(AbstractPrinter):
+    """
+    A per-function pretty-printer of vmprof profiles.
+    """
+
+    def __init__(self, include_callees, no_native, percent_cutoff):
+        self.include_callees = include_callees
+        self.no_native = no_native
+        self.percent_cutoff = percent_cutoff
+
+    def _show(self, tree):
+        self._print_tree(tree)
+
+    def _walk_tree(self, parent, node, callback):
+        callback(parent, node)
+        for c in six.itervalues(node.children):
+            self._walk_tree(node, c, callback)
+
+    def _print_tree(self, tree):
+        func_id_to_count = {}
+
+        def collect_node(parent, node):
+            ndescr = parse_block_name(node.name)
+            if self.no_native and ndescr.block_type == 'n':
+                return
+
+            mycount = node.count
+            if not self.include_callees:
+                mycount = mycount - sum(
+                        0 if parse_block_name(ch.name)[0] == 'n' and self.no_native
+                        else ch.count
+
+                    for ch in six.itervalues(node.children))
+
+            func_id_to_count[ndescr] = func_id_to_count.get(ndescr, 0) + mycount
+
+        self._walk_tree(None, tree, collect_node)
+
+        cost_list = sorted(
+                func_id_to_count.items(),
+                key=(lambda item: item[1]),
+                reverse=True)
+
+        total = float(tree.count)
+
+        for ndescr, count in cost_list:
+            percent = count/total*100
+
+            if percent >= self.percent_cutoff:
+                print(
+                        f'{percent:10.3f}% - '
+                        f'{color(ndescr.funname, color.WHITE, bold=True)}:'
+                        f'{ndescr.filename}:{ndescr.funline}')
```

Comment:
We can't accept f-strings until we drop python < 3.6

Reply:
d950361